uaclient/tests/test_ua_update_messaging.py

```diff
@@ -0,0 +1,427 @@
+import datetime
+import mock
+import os
+
+import pytest
+
+from uaclient.defaults import (
+    BASE_ESM_URL,
+    BASE_UA_URL,
+    CONTRACT_EXPIRY_GRACE_PERIOD_DAYS,
+)
+from uaclient.status import (
+    MESSAGE_ANNOUNCE_ESM,
+    MESSAGE_CONTRACT_EXPIRED_APT_NO_PKGS_TMPL,
+    MESSAGE_CONTRACT_EXPIRED_APT_PKGS_TMPL,
+    MESSAGE_CONTRACT_EXPIRED_GRACE_PERIOD_TMPL,
+    MESSAGE_CONTRACT_EXPIRED_SOON_TMPL,
+    MESSAGE_DISABLED_MOTD_NO_PKGS_TMPL,
+    MESSAGE_DISABLED_APT_PKGS_TMPL,
+    MESSAGE_UBUNTU_NO_WARRANTY,
+    ApplicationStatus,
+)
+from uaclient import util
+
+from lib.ua_update_messaging import (
+    ContractExpiryStatus,
+    ExternalMessage,
+    get_contract_expiry_status,
+    update_apt_and_motd_messages,
+    write_apt_and_motd_templates,
+    write_esm_announcement_message,
+    _write_esm_service_msg_templates,
+)
+
+M_PATH = "lib.ua_update_messaging."
+
+
+class TestGetContractExpiryStatus:
+    @pytest.mark.parametrize(
+        "contract_remaining_days,expected_status",
+        (
+            (21, ContractExpiryStatus.ACTIVE),
+            (20, ContractExpiryStatus.ACTIVE_EXPIRED_SOON),
+            (-1, ContractExpiryStatus.EXPIRED_GRACE_PERIOD),
```

Comment:
We can add in some more edges here pretty easily
```suggestion
            (20, ContractExpiryStatus.ACTIVE_EXPIRED_SOON),
            (0, ContractExpiryStatus.ACTIVE_EXPIRED_SOON),
            (-1, ContractExpiryStatus.EXPIRED_GRACE_PERIOD),
            (-14, ContractExpiryStatus.EXPIRED_GRACE_PERIOD),
            (-15, ContractExpiryStatus.EXPIRED),
```